sys/parsoid.js

```diff
@@ -264,18 +264,44 @@ PSP.generateAndSave = function(hyper, req, format, currentContentRes) {
             hyper.metrics.increment('sys_parsoid_generateAndSave.unchanged_rev_render');
             return currentContentRes;
         } else if (res.status === 200) {
+
             var resp = {
                 status: res.status,
                 headers: res.body[format].headers,
                 body: res.body[format].body
             };
+
+            var redirectUpdate = P.resolve();
+            // Extract redirect target, if any
+            var redirectTarget = mwUtil.extractRedirect(res.body.html.body);
+            if (redirectTarget) {
+                // This revision is actually a redirect. Pass redirect target
+                // to caller, and let it rewrite the location header.
+                resp.status = 302;
+                resp.headers.location = redirectTarget;
+                // Also save the location header.
+                res.body.html.headers.location = redirectTarget;
+                redirectUpdate = hyper.post({
+                    uri: new URI([rp.domain, 'sys', 'page_revisions',
+                                 'restrictions', rp.title, rp.revision]),
+                    body: {
+                        redirect: redirectTarget,
+                    }
+                })
+                .catch(function(e) {
+                    hyper.log('error/parsoid/redirect/update', e);
+                    throw e;
+                });
+            }
+
             resp.headers.etag = mwUtil.makeETag(rp.revision, tid);
             return self.saveParsoidResult(hyper, req, format, tid, res)
             .then(function() {
                 var dependencyUpdate = self._dependenciesUpdate(hyper, req);
                 if (mwUtil.isNoCacheRequest(req)) {
                     // Finish background updates before returning
-                    return dependencyUpdate.thenReturn(resp);
+                    return P.all([redirectUpdate, dependencyUpdate])
```

Comment:
I'm wondering if we need to update redirects sync before dependencies. The dependency update might use the redirect information (to not access mobile service to generate useless redirect content)

Although it's a minor issue as we will remove dependency updates soon enough